### ISSUE TITLE
Use of Map for session add/lookup

### ISSUE THREAD

**Issue by krsna1729** [wmnsk/go-gtp#62]
I was trying to benchmark the current implementation of adding session and a map. The performance of the current approach of iterating over the sessions seems low. Would it be better to maintain `map[teid]*Session` and map[imsi]*Session`? Let me know if what/how I am benchmarking is wrong.

```    
func benchmarkAddSession(numExisitingSessions int, b *testing.B) {
    benchConn := &v2.Conn{Sessions: []*v2.Session{}}
    for i := 0; i < numExisitingSessions; i++ {
        imsi := fmt.Sprintf("%015d", i)
        benchConn.AddSession(v2.NewSession(dummyAddr, &v2.Subscriber{IMSI: imsi}))
    }
    b.ResetTimer()
    for i := 1; i <= b.N; i++ {
        benchConn.AddSession(v2.NewSession(dummyAddr, &v2.Subscriber{IMSI: "001011234567891"}))
    }
}
func BenchmarkAddSessionExist0(b *testing.B)   { benchmarkAddSession(0, b) }
func BenchmarkAddSessionExist100(b *testing.B) { benchmarkAddSession(1e2, b) }
func BenchmarkAddSessionExist1K(b *testing.B)  { benchmarkAddSession(1e3, b) }
func BenchmarkAddSessionExist10K(b *testing.B) { benchmarkAddSession(1e4, b) }
```
```
func benchmarkAddIMSISessionMap(numExisitingSessions int, b *testing.B) {
    IMSISession := make(map[string]*v2.Session)
    for i := 0; i < numExisitingSessions; i++ {
        imsi := fmt.Sprintf("%015d", i)
        IMSISession[imsi] = v2.NewSession(dummyAddr, &v2.Subscriber{IMSI: imsi})
    }
    b.ResetTimer()
    for i := 1; i <= b.N; i++ {
        IMSISession["001011234567891"] = v2.NewSession(dummyAddr, &v2.Subscriber{IMSI: "001011234567891"})
    }
}
func BenchmarkAddIMSISessionMapExist0(b *testing.B)    { benchmarkAddIMSISessionMap(0, b) }
func BenchmarkAddIMSISessionMapExist100(b *testing.B)  { benchmarkAddIMSISessionMap(1e2, b) }
func BenchmarkAddIMSISessionMapExist1K(b *testing.B)   { benchmarkAddIMSISessionMap(1e3, b) }
func BenchmarkAddIMSISessionMapExist10K(b *testing.B)  { benchmarkAddIMSISessionMap(1e4, b) }
func BenchmarkAddIMSISessionMapExist100K(b *testing.B) { benchmarkAddIMSISessionMap(1e5, b) }
func BenchmarkAddIMSISessionMapExist1M(b *testing.B)   { benchmarkAddIMSISessionMap(1e6, b) }
```
```
func benchmarkAddIMSISessionSyncMap(numExisitingSessions int, b *testing.B) {
    var IMSISession sync.Map
    for i := 0; i < numExisitingSessions; i++ {
        imsi := fmt.Sprintf("%015d", i)
        IMSISession.Store(imsi, v2.NewSession(dummyAddr, &v2.Subscriber{IMSI: imsi}))
    }
    b.ResetTimer()
    for i := 1; i <= b.N; i++ {
        IMSISession.Store("001011234567891", v2.NewSession(dummyAddr, &v2.Subscriber{IMSI: "001011234567891"}))
    }
}
func BenchmarkAddIMSISessionSyncMapExist0(b *testing.B)    { benchmarkAddIMSISessionSyncMap(0, b) }
func BenchmarkAddIMSISessionSyncMapExist100(b *testing.B)  { benchmarkAddIMSISessionSyncMap(1e2, b) }
func BenchmarkAddIMSISessionSyncMapExist1K(b *testing.B)   { benchmarkAddIMSISessionSyncMap(1e3, b) }
func BenchmarkAddIMSISessionSyncMapExist10K(b *testing.B)  { benchmarkAddIMSISessionSyncMap(1e4, b) }
func BenchmarkAddIMSISessionSyncMapExist100K(b *testing.B) { benchmarkAddIMSISessionSyncMap(1e5, b) }
func BenchmarkAddIMSISessionSyncMapExist1M(b *testing.B)   { benchmarkAddIMSISessionSyncMap(1e6, b) }
```

```
BenchmarkAddSessionExist0-88                       99534         11564 ns/op       17216 B/op         19 allocs/op
BenchmarkAddSessionExist100-88                     62908         18471 ns/op       19248 B/op         26 allocs/op
BenchmarkAddSessionExist1K-88                      28222         42054 ns/op       33584 B/op         29 allocs/op
BenchmarkAddSessionExist10K-88                      3542        401157 ns/op      403505 B/op         38 allocs/op
<too long to test>

BenchmarkAddIMSISessionMapExist0-88                84746         14651 ns/op       17208 B/op         18 allocs/op
BenchmarkAddIMSISessionMapExist100-88              81644         15163 ns/op       17208 B/op         18 allocs/op
BenchmarkAddIMSISessionMapExist1K-88               95305         12134 ns/op       17208 B/op         18 allocs/op
BenchmarkAddIMSISessionMapExist10K-88             114638          8823 ns/op       17208 B/op         18 allocs/op
BenchmarkAddIMSISessionMapExist100K-88            380025          7923 ns/op       17208 B/op         18 allocs/op
BenchmarkAddIMSISessionMapExist1M-88              388423          7936 ns/op       17208 B/op         18 allocs/op

BenchmarkAddIMSISessionSyncMapExist0-88            80818         15798 ns/op       17224 B/op         19 allocs/op
BenchmarkAddIMSISessionSyncMapExist100-88          73264         15957 ns/op       17224 B/op         19 allocs/op
BenchmarkAddIMSISessionSyncMapExist1K-88           93626         12520 ns/op       17224 B/op         19 allocs/op
BenchmarkAddIMSISessionSyncMapExist10K-88         109438          9830 ns/op       17224 B/op         19 allocs/op
BenchmarkAddIMSISessionSyncMapExist100K-88        190755          9579 ns/op       17224 B/op         19 allocs/op
BenchmarkAddIMSISessionSyncMapExist1M-88          121707          8270 ns/op       17224 B/op         19 allocs/op
```

**Comment by wmnsk** [comment]
Thank you for benchmarking!

Your benchmark looks good, and according to the results it seems we should follow your idea. FYI, the reason why I use `sync.Map` is that the sessions might be retrieved often in their lifetimes. So if it’s slower at the time it’s added, I thought using `sync.Map` was better. For sure it depends on the usage(if a session is simply created and not read/updated before it’s deleted, default `map` seems to be better), and nothing’s going to be better than benchmarking that, too, though... 😅

https://golang.org/pkg/sync/#Map

So, if you’d be willing, could you work on that and send PR?

**Comment by krsna1729** [comment]
I will send a PR this week hopefully.

**Comment by wmnsk** [comment]
Looking forward to that. Thank you 😃

**Comment by wmnsk** [comment]
Fixed in #67. Thanks @krsna1729!
